provider/azure.go

```diff
@@ -0,0 +1,456 @@
+	} else {
+		environment, err = azure.EnvironmentFromName(cfg.Cloud)
+		if err != nil {
+			return nil, fmt.Errorf("invalid cloud value '%s': %v", cfg.Cloud, err)
+		}
+	}
+
+	oauthConfig, err := adal.NewOAuthConfig(environment.ActiveDirectoryEndpoint, cfg.TenantID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retreive OAuth config: %v", err)
+	}
+
+	token, err := adal.NewServicePrincipalToken(*oauthConfig, cfg.ClientID, cfg.ClientSecret, environment.ResourceManagerEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create service principal token: %v", err)
+	}
+
+	zonesClient := dns.NewZonesClient(cfg.SubscriptionID)
+	zonesClient.Authorizer = autorest.NewBearerAuthorizer(token)
+	recordsClient := dns.NewRecordSetsClient(cfg.SubscriptionID)
+	recordsClient.Authorizer = autorest.NewBearerAuthorizer(token)
+
+	provider := &AzureProvider{
+		domainFilter:  domainFilter,
+		dryRun:        dryRun,
+		resourceGroup: cfg.ResourceGroup,
+		zonesClient:   zonesClient,
+		recordsClient: recordsClient,
+	}
+	return provider, nil
+}
+
+// Records gets the current records.
+//
+// Returns the current records or an error if the operation failed.
+func (p *AzureProvider) Records() (endpoints []*endpoint.Endpoint, _ error) {
+	zones, err := p.azureZones()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, zone := range zones {
+		err := p.iterateRecords(*zone.Name, func(recordSet dns.RecordSet) bool {
+			if recordSet.Name == nil || recordSet.Type == nil {
+				log.Error("Skipping invalid record set with nil name or type.")
+				return true
+			}
+			recordType := strings.TrimLeft(*recordSet.Type, "Microsoft.Network/dnszones/")
+			switch dns.RecordType(recordType) {
+			case dns.A, dns.CNAME, dns.TXT:
+				name := formatDNSName(*recordSet.Name, *zone.Name)
+				target := extractTarget(recordType, &recordSet)
+				if target == "" {
+					log.Errorf("Failed to extract target for '%s' with type '%s'.", name, recordType)
+					return true
+				}
+				endpoint := endpoint.NewEndpoint(name, target, recordType)
+				log.Debugf(
+					"Adding %s record for '%s' with target '%s'.",
+					endpoint.RecordType,
+					endpoint.DNSName,
+					endpoint.Target,
+				)
+				endpoints = append(endpoints, endpoint)
+			default:
+			}
+			return true
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+	return endpoints, nil
+}
+
+// ApplyChanges applies the given changes.
+//
+// Returns nil if the operation was successful or an error if the operation failed.
+func (p *AzureProvider) ApplyChanges(changes *plan.Changes) error {
+	zones, err := p.azureZones()
+	if err != nil {
+		return err
+	}
+
+	deleted, updated := mapAzureChanges(zones, changes)
+	p.deleteAzureRecords(deleted)
+	p.updateAzureRecords(updated)
+	return nil
+}
+
+func (p *AzureProvider) azureZones() ([]dns.Zone, error) {
+	log.Debug("Retrieving Azure DNS zones.")
+
+	var zones []dns.Zone
+	list, err := p.zonesClient.ListByResourceGroup(p.resourceGroup, nil)
```

Comment:
```
// ListByResourceGroup lists the DNS zones within a resource group.
//
// resourceGroupName is the name of the resource group. top is the maximum
// number of record sets to return. If not specified, returns up to 100 record
// sets.
```
This is taken from azure sdk source code. Correct me if I am wrong but one should use `ListByResourceGroupNextResults` to get the full list ?